cmd/entrypoint/webhooks.go

```diff
@@ -0,0 +1,312 @@
+	webhook := &conversion.Webhook{}
+	if err := webhook.InjectScheme(scheme); err != nil {
+		return err
+	}
+	mux := http.NewServeMux()
+	mux.HandleFunc(webhookPath, webhook.ServeHTTP)
+
+	// need a k8s client
+	kubeinfo := k8s.NewKubeInfo("", "", "")
+	restConfig, err := kubeinfo.GetRestConfig()
+	if err != nil {
+		return err
+	}
+	coreClient, err := k8sClientCoreV1.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	// get CA secret
+	secrets := coreClient.Secrets(GetAmbassadorNamespace())
+	caSecret, err := secrets.Get(ctx, caSecretName, k8sTypesMetaV1.GetOptions{})
+	var caPEM *b.Buffer
+	var caTemplate *x509.Certificate
+	var caPrivKey *rsa.PrivateKey
+	if err != nil {
+		// Error here most likely means the secret doesnt exist yet
+		// (then we make one). Otherwise if its an actual error we exit here
+		if !k8sErrors.IsNotFound(err) {
+			return err
+		}
+
+		// Generate CA Certificate and key...
+		notBefore := time.Now()
+		notAfter := notBefore.Add(time.Duration(certValidDays*24) * time.Hour)
+		serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+		serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+		if err != nil {
+			return err
+		}
+		caTemplate = &x509.Certificate{
+			SerialNumber: serialNumber,
+			Subject: pkix.Name{
+				Organization: []string{"Ambassador Labs"},
+			},
+			NotBefore:             notBefore,
+			NotAfter:              notAfter,
+			IsCA:                  true,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+			KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+			BasicConstraintsValid: true,
+		}
+		caPrivKey, err = rsa.GenerateKey(rand.Reader, 4096)
+		if err != nil {
+			return err
+		}
+		caBytes, err := x509.CreateCertificate(rand.Reader, caTemplate, caTemplate, &caPrivKey.PublicKey, caPrivKey)
+		if err != nil {
+			return err
+		}
+
+		// PEM encode certificate and key
+		caPEM = new(b.Buffer)
+		if err := pem.Encode(caPEM, &pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: caBytes,
+		}); err != nil {
+			return err
+		}
+		caPrivKeyPEM := new(b.Buffer)
+		if err := pem.Encode(caPrivKeyPEM, &pem.Block{
+			Type:  "RSA PRIVATE KEY",
+			Bytes: x509.MarshalPKCS1PrivateKey(caPrivKey),
+		}); err != nil {
+			return err
+		}
+
+		// Create and write the secret
+		_, err = secrets.Create(ctx, &k8sTypesCoreV1.Secret{
+			ObjectMeta: k8sTypesMetaV1.ObjectMeta{
+				Name:      caSecretName,
+				Namespace: GetAmbassadorNamespace(),
+			},
+			Type: k8sTypesCoreV1.SecretTypeTLS,
+			Data: map[string][]byte{
+				k8sTypesCoreV1.TLSPrivateKeyKey: caPrivKeyPEM.Bytes(),
+				k8sTypesCoreV1.TLSCertKey:       caPEM.Bytes(),
+			},
+		}, k8sTypesMetaV1.CreateOptions{})
+		if err != nil {
+			if k8sErrors.IsAlreadyExists(err) {
+				caSecret, err = secrets.Get(ctx, caSecretName, k8sTypesMetaV1.GetOptions{})
+				if err != nil {
+					return err
+				}
+			} else {
+				return err
+			}
+		}
+	}
```

Comment:
Unless I'm reading this wrong, `caSecret` ends up `nil` in the case where it actually creates the secret.